File: java/Tree/avl.java
package tree;

class Node {
    int val;
    Node left;
    Node right;
    int ht;
    Node (int x){
        this.val = x;
        this.left = this.right = null;
        this.ht = 1;
    }
}
class avlTree {
    Node Root = null;
    public int height(Node n){
        if(n == null) return 0;
        return n.ht;
    }

    public int balanceFactor(Node n){
        if(n == null) return 0;
        return height(n.left) - height(n.right);
    }

    public Node rightrotate(Node n){
        Node x = n.left;
        Node t2 = x.right;

        x.right = n;
        n.left = t2;

        n.ht = 1 + Math.max(height(n.left), height(n.right));
        x.ht = 1 + Math.max(height(x.left), height(x.right));
        
        return x;
    }

    public Node leftrotate(Node n){
        Node x = n.right;
        Node t2 = x.left;

        x.left = n;
        n.right = t2;

        n.ht = 1 + Math.max(height(n.left), height(n.right));
        x.ht = 1 + Math.max(height(x.left), height(x.right));
        return x;
    }

    public Node insert(Node n, int key){
        if(n == null) {
            this.Root = new Node(key);
            return this.Root;
        }
        if(key < this.Root.val){
            n.left = insert(n.left, key);
        } else if (key > this.Root.val){
            n.right = insert(n.right, key);
        } else {
            return n;
        }

        n.ht = 1 + Math.max(height(n.left), height(n.right));
        int bf = balanceFactor(n);
        // Rotation wala part
        if (bf > 1 && key < n.left.val)
            return rightrotate(n);

        // Right Right Case (RR Rotation)
        if (bf < -1 && key > n.right.val)
            return leftrotate(n);

        // Left Right Case (LR Rotation)
        if (bf > 1 && key > n.left.val) {
            n.left = leftrotate(n.left);
            return rightrotate(n);
        }

        // Right Left Case (RL Rotation)
        if (bf < -1 && key < n.right.val) {
            n.right = rightrotate(n.right);
            return leftrotate(n);
        }

        return n;
    }

    public void dis(Node r){
        if(r == null) return;
        dis(r.left);
        System.out.print(r.val + " ");
        dis(r.right);
    }

}



public class avl {
    public static void main(String args[]){

        avlTree at = new avlTree();
        at.insert(at.Root, 1);
        at.insert(at.Root, 3);
        at.insert(at.Root, 5);
        at.insert(at.Root, 2);

        at.dis(at.Root);
    }
}

File: python/Tree/avl.py
class Node:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None
        self.height = 1  # New node has height 1


class AVL:
    def __init__(self):
        self.root = None  # Root of the AVL Tree

    def height(self, node):
        return node.height if node else 0

    def balance(self, node):
        return self.height(node.left) - self.height(node.right) if node else 0

    def leftrotate(self, z):
        y = z.right
        T2 = y.left

        y.left = z
        z.right = T2

        z.height = max(self.height(z.left), self.height(z.right)) + 1
        y.height = max(self.height(y.left), self.height(y.right)) + 1

        return y

    def rightrotate(self, z):
        y = z.left
        T3 = y.right

        y.right = z
        z.left = T3

        z.height = max(self.height(z.left), self.height(z.right)) + 1
        y.height = max(self.height(y.left), self.height(y.right)) + 1

        return y

    def insert(self, root, key):
        if root is None:
            return Node(key)

        if key < root.val:
            root.left = self.insert(root.left, key)
        elif key > root.val:
            root.right = self.insert(root.right, key)
        else:
            return root  # No duplicates allowed

        # Update height of the current node
        root.height = max(self.height(root.left), self.height(root.right)) + 1

        # Get balance factor
        bf = self.balance(root)

        # Left Heavy (Right Rotation)
        if bf > 1:
            if key < root.left.val:  # Left-Left Case
                return self.rightrotate(root)
            else:  # Left-Right Case
                root.left = self.leftrotate(root.left)
                return self.rightrotate(root)

        # Right Heavy (Left Rotation)
        if bf < -1:
            if key > root.right.val:  # Right-Right Case
                return self.leftrotate(root)
            else:  # Right-Left Case
                root.right = self.rightrotate(root.right)
                return self.leftrotate(root)

        return root

    def insert_key(self, key):
        self.root = self.insert(self.root, key)

    def min_value_node(self, node):
        """Finds the node with the minimum value (leftmost node)"""
        while node.left:
            node = node.left
        return node

    def delete(self, root, key):
        if root is None:
            return root

        # Standard BST deletion
        if key < root.val:
            root.left = self.delete(root.left, key)
        elif key > root.val:
            root.right = self.delete(root.right, key)
        else:
            # Node with one child or no child
            if root.left is None:
                return root.right
            elif root.right is None:
                return root.left

            # Node with two children: Get inorder successor (smallest in right subtree)
            temp = self.min_value_node(root.right)
            root.val = temp.val  # Copy successor value to current node
            root.right = self.delete(root.right, temp.val)  # Delete successor

        if root is None:
            return root

        # Update height
        root.height = max(self.height(root.left), self.height(root.right)) + 1

        # Get balance factor
        bf = self.balance(root)

        # Left Heavy
        if bf > 1:
            if self.balance(root.left) >= 0:  # Left-Left Case
                return self.rightrotate(root)
            else:  # Left-Right Case
                root.left = self.leftrotate(root.left)
                return self.rightrotate(root)

        # Right Heavy
        if bf < -1:
            if self.balance(root.right) <= 0:  # Right-Right Case
                return self.leftrotate(root)
            else:  # Right-Left Case
                root.right = self.rightrotate(root.right)
                return self.leftrotate(root)

        return root

    def delete_key(self, key):
        self.root = self.delete(self.root, key)

    def levelorder(self, root):
        if not root:
            return
        queue = [root]

        while queue:
            level_size = len(queue)
            for _ in range(level_size):
                node = queue.pop(0)
                print(node.val, end=" ")

                if node.left:
                    queue.append(node.left)
                if node.right:
                    queue.append(node.right)

            print()  # New line for each level


# Driver Code
avl_tree = AVL()

for i in range(1, 8):  # Inserting values 1 to 7
    avl_tree.insert_key(i)

print("Level Order Traversal of AVL Tree:")
avl_tree.levelorder(avl_tree.root)

print("\nDeleting node 4:")
avl_tree.delete_key(4)
avl_tree.levelorder(avl_tree.root)

print("\nDeleting node 2:")
avl_tree.delete_key(2)
avl_tree.levelorder(avl_tree.root)
